db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/PredicateList.java

```diff
@@ -773,41 +847,224 @@ private void orderUsefulPredicates(Optimizable optTable,
+          if (inlistPreds.size() > 1) {
+            predsForNewInList = new ArrayList<>();
+            int firstPred = -1, lastPred = -1, lastIndexPos = -1, firstInListPred = -1;
+
+            boolean foundPred[] = new boolean[usefulCount];
+            int numConstants = 1;
+            int maxMulticolumnProbeValues = getCompilerContext().getMaxMulticolumnProbeValues();
+            for (int i = 0; i < usefulCount; i++) {
+                final Predicate pred = usefulPredicates[i];
+
+                if (pred.getIndexPosition() == lastIndexPos + 1) {
+                    BinaryRelationalOperatorNode bron =
+                        (BinaryRelationalOperatorNode) pred.getRelop();
+                    boolean inList = pred.isInListProbePredicate();
+                    if (inList ||
+                        (bron != null && bron.getOperator() == RelationalOperator.EQUALS_RELOP)) {
+                        if (inList) {
+                            InListOperatorNode inNode = pred.getSourceInList(true);
+    
+                            if (firstInListPred == -1)
+                                firstInListPred = i;
+                            
+                            // A max setting of -1 means there is no limit to the size of
+                            // IN list that we generate.
+                            if (maxMulticolumnProbeValues != -1 &&
+                                (numConstants * inNode.rightOperandList.size()) > maxMulticolumnProbeValues)
+                                break;
+                            numConstants *= inNode.rightOperandList.size();
+                        }
+                        if (firstPred == -1)
+                            firstPred = i;
+                        lastPred = i;
+                        predsForNewInList.add(usefulPredicates[i]);
+                        lastIndexPos = pred.getIndexPosition();
+                        foundPred[i] = true;
+                    }
+                }
+                // Can't have any gaps in index position.
+                if (pred.getIndexPosition() > lastIndexPos+1)
+                    break;
+            }
 
-                for (int pos = 0; pos < inlistPosition; pos++) {
-                    if (!isEquality[pos]) {
-                        inlistQualified = false;
-                        break;
+            ValueNodeList vnl = (ValueNodeList)getNodeFactory().getNode(
+                                 C_NodeTypes.VALUE_NODE_LIST,
+                                  getContextManager());
+            ValueNodeList groupedConstants =
+                   (ValueNodeList) getNodeFactory().getNode(
+                                   C_NodeTypes.VALUE_NODE_LIST,
+                                   getContextManager());
+            boolean multiColumnInListBuilt = false;
+            if (optTable instanceof FromBaseTable &&
+                (!((FromBaseTable) optTable).isSpark(((FromBaseTable) optTable).getDataSetProcessorType()) ||
```

Comment:
The decision of whether to go for Spark path or not does not depend on one step. It is possible that for this FromBaseTable node, we don't choose spark path, but another FromBaseTable node would choose spark path. The final decision is made after all the steps in the plan have been checked. The check here may not block the multi-probe scan completely for multiple inlists for Spark path.

Reply:
The dataset processor type finally used is the one tracked in the current CompilerContext.dataSetProcessorType. During the code generation phase, as we go through each operation, this variable will be updated. Please see the call in FromBaseTable.getScanArguements()->acb.setDataSetProcessorType(dataSetProcessorType);

Reply:
Opened [SPLICE-2289](https://splice.atlassian.net/browse/SPLICE-2289) for this issue.